h/services/group.py

```diff
@@ -96,19 +96,38 @@ def create_restricted_group(self, name, userid, origins, description=None, organ
                             organization=organization,
                             )
 
-    def update_membership(self, group, usernames):
-        group_member_usernames = [member.username for member in group.members]
+    def add_members(self, group, userids):
+        """
+        Add the users indicated by userids to this group's members.
+
+        Any pre-existing members will not be affected.
+        """
+        for userid in userids:
+            self.member_join(group, userid)
 
-        usernames_to_add = list(set(usernames) - set(group_member_usernames))
-        usernames_to_remove = list(set(group_member_usernames) - set(usernames))
+    def update_members(self, group, userids):
+        """
+        Update this group's membership to be the list of users indicated by
+        userids.
+
+        The users indicated by userids will *replace* the members of this group.
+        Any pre-existing member whose userid is not present in userids will
+        be removed as a member.
+
+        :param group:   group model
+        :param userids: the list of userids corresponding to users who should
+                        be the members of this group
+        """
+        current_member_userids = [member.userid for member in group.members]
+        userids_for_removal = list(
```

Comment:
Could this have just been another list comprehension?

```python
userids_for_removal = [userid for userid in current_member_userids if userid not in userids]
```

This is the generally preferred style for doing this in modern Python, unless I'm missing a reason why `filter` or `set` is needed here

Reply:
This actually produces a warning from pylint:

```
pylint: deprecated-lambda / map/filter on lambda could be replaced by comprehension
```

Weirdly pylint gives this warning only in Python 2 and not 3, but I'm not sure why that is.

Reply:
I googled and couldn't find confirmation but at a guess I would say that pylint's `deprecated-lambda` warning is never given in Python 3 because `filter` returns an iterable rather than a list in Python 3 so technically a list comprehension is _not_ equivalent to a `filter` in 3, though it is in 2.

In most cases though, including I think this one, a list comprehension will do just fine instead of `filter` in Py3 as well as 2.

(P.S. Here is [Guido saying that comprehensions are preferred over `filter` or `map`](https://www.artima.com/weblogs/viewpost.jsp?thread=98196), which is what led to `deprecated-lambda` being added to pylint.)

Reply:
Totally an oversight on my end. I had been doing something more complex which called for a filter at one point but this could be a comprehension and I shall fix.